pic-sure-auth-services/src/main/java/edu/harvard/hms/dbmi/avillach/auth/service/auth/FENCEAuthenticationService.java

```diff
@@ -159,108 +176,173 @@ public Response getFENCEProfile(String callback_url, Map<String, String> authReq
+                continue;
+            }
+
+            String projectId = studyMetadata.getStudy_identifier();
+            String consentCode = studyMetadata.getConsent_group_code();
+            String newRoleName = StringUtils.isNotBlank(consentCode) ? "FENCE_"+projectId+"_"+consentCode : "FENCE_"+projectId;
+
+            project_access_set.add(newRoleName);
         }
 
-        final String idp = extractIdp(current_user);
+        // Project access set is now a set of role names that should be assigned to the user
+        logger.info("getFENCEProfile() project access set: {}", project_access_set);
+
+        // Step 1: Remove roles that are not in the project_access_set
+        Set<Role> rolesToRemove = new HashSet<>();
+        // Also, track the roles that are assigned to the user and in the project_access_set
+        Set<String> rolesAssigned = new HashSet<>();
+        for (Role role : current_user.getRoles()) {
+            if (!project_access_set.contains(role.getName())
+                    && !role.getName().startsWith("MANUAL_")
+                    && !role.getName().equals(fence_open_access_role_name)
+                    && !role.getName().equals("PIC-SURE Top Admin")
+                    && !role.getName().equals("Admin")) {
+                rolesToRemove.add(role);
+            }
 
-        if (current_user.getRoles() != null && (current_user.getRoles().size() > 0 || openAccessIdpValues.contains(idp))) {
+            if (project_access_set.contains(role.getName())) {
+                rolesAssigned.add(role.getName());
+            }
+        }
+
+        // Remove roles that are not in the project_access_set
+        if (!rolesToRemove.isEmpty()) {
+            current_user.getRoles().removeAll(rolesToRemove);
+        }
+
+        // Given the set of roles assigned and that set of roles that should be assigned, we can reduce the set of roles from the project_access_set
+        // to only those that are not in the rolesAssigned set
+        if (!rolesAssigned.isEmpty()) {
+            logger.info("getFENCEProfile() roles that are assigned: {}", rolesAssigned);
+            project_access_set.removeAll(rolesAssigned);
+        }
+
+        if (!project_access_set.isEmpty()) {
+            logger.info("getFENCEProfile() roles that should be assigned: {}", project_access_set);
+
+            // Given our reduced list of roles that should be assigned, we can determine which of those roles are not in the database
+            // This also tells use which roles are in the database
+            Set<String> rolesThatExist = roleRepo.getRoleNamesByNames(project_access_set);
+            if (!rolesThatExist.isEmpty()) {
+                // Assign the roles that exist in the database to the user
+                logger.info("getFENCEProfile() assigning roles that exist in the database: {}", rolesThatExist);
+                Set<Role> rolesThatExistSet = roleRepo.getRolesByNames(rolesThatExist);
```

Comment:
I'm confused. Why not just get the full role objects in the first query?

Reply:
For this code I have two options:
Current:
```
Set<String> rolesThatExist = roleRepo.getRoleNamesByNames(project_access_set);
if (!rolesThatExist.isEmpty()) {
    Set<Role> rolesThatExistSet = roleRepo.getRolesByNames(rolesThatExist);
    current_user.getRoles().addAll(rolesThatExistSet);

    project_access_set.removeAll(rolesThatExist);
} 
```
or
```
Set<Role> rolesThatExist = roleRepo.getRolesByNames(project_access_set);
if (!rolesThatExist.isEmpty()) {
    current_user.getRoles().addAll(rolesThatExist);

    Set<String> rolesThatExistNames = rolesThatExist.stream().map(Role::getName).collect(Collectors.toSet());
    project_access_set.removeAll(rolesThatExistNames);
}
```
The second is likely faster. I don't know how much faster. I will test.